File: windows/Sources/TutorialWindow.cs
using System.Drawing.Drawing2D;

namespace CaffeineBar;

/// A short, self-contained walkthrough shown on first launch and re-openable
/// anytime from the tray's right-click menu.
///
/// Windows tucks newly-registered tray icons into the overflow flyout, so a
/// first-time user can easily install this and never see the icon at all. Step
/// two exists specifically to solve that; the rest explain the three
/// interactions (click, hover, right-click) that aren't discoverable on their own.
///
/// Unlike CustomMenuWindow this is a normal, focusable window — it's a
/// destination the user reads and clicks through, not a transient panel.
public sealed class TutorialWindow : Form
{
    private sealed record Step(string Heading, string Body);

    private static readonly Step[] Steps =
    [
        new("Keep your PC awake",
            "Caffeine Bar holds a display-awake request for as long as it's active, so your "
            + "screen won't sleep and your session stays alive. Switch it off and your normal "
            + "power plan applies again — nothing is changed permanently."),

        new("Find the icon",
            "Windows hides newly-installed tray icons, so Caffeine Bar starts out behind the "
            + "˄ arrow next to the clock.\n\nTo keep it visible, open Settings › "
            + "Personalization › Taskbar › Other system tray icons and switch "
            + "Caffeine Bar on. Dragging it out of the ˄ flyout does the same thing."),

        new("Click to toggle",
            "Click the icon to switch it on and off.\n\nFilled means active — the can fills in "
            + "and sprouts wings, or the cup puffs steam, depending on your icon style. An "
            + "empty outline means it's off."),

        new("Hover for the menu",
            "Hover the icon to open the full panel: how long you've been active, the energy "
            + "drink and coffee finders, and the icon style switch.\n\nPick Coffee or Energy "
            + "Drink under Icon Style to change which drink the icon shows."),

        new("Right-click for settings",
            "Right-click for Launch at Login, Prevent Sleep on Lid Close, this tutorial, and Quit."
            + "\n\nLid close is a hardware sleep trigger, so it needs admin rights once. Caffeine "
            + "Bar registers two scheduled tasks that first time and never has to ask again."),
    ];

    private int _index;
    private readonly PillButton _backButton;
    private readonly PillButton _nextButton;
    private readonly PillButton _closeButton;

    private const int WindowWidth = 460;
    private const int WindowHeight = 320;
    private const int Pad = 26;
    private const int CornerRadius = 16;

    public TutorialWindow()
    {
        FormBorderStyle = FormBorderStyle.None;
        StartPosition = FormStartPosition.CenterScreen;
        ShowInTaskbar = true;
        Text = "Caffeine Bar Tutorial";
        BackColor = Palette.Panel;
        ClientSize = new Size(WindowWidth, WindowHeight);
        DoubleBuffered = true;
        KeyPreview = true;
        SetStyle(ControlStyles.AllPaintingInWmPaint | ControlStyles.UserPaint |
                 ControlStyles.OptimizedDoubleBuffer | ControlStyles.ResizeRedraw, true);

        _backButton = new PillButton("Back", Palette.White(0.55), filled: false)
        {
            Left = Pad,
            Top = WindowHeight - Pad - PillButton.ButtonHeight,
        };
        _backButton.Clicked += () => Go(_index - 1);

        _nextButton = new PillButton("Next", Palette.Energy, filled: true)
        {
            Top = WindowHeight - Pad - PillButton.ButtonHeight,
        };
        _nextButton.Clicked += () =>
        {
            if (_index == Steps.Length - 1) Close();
            else Go(_index + 1);
        };

        _closeButton = new PillButton("Skip", Palette.White(0.4), filled: false)
        {
            Left = Pad,
            Top = WindowHeight - Pad - PillButton.ButtonHeight,
        };
        _closeButton.Clicked += Close;

        Controls.Add(_backButton);
        Controls.Add(_nextButton);
        Controls.Add(_closeButton);
        Go(0);
    }

    protected override void OnHandleCreated(EventArgs e)
    {
        base.OnHandleCreated(e);
        int preference = Native.DWMWCP_ROUND;
        if (Native.DwmSetWindowAttribute(Handle, Native.DWMWA_WINDOW_CORNER_PREFERENCE,
                ref preference, sizeof(int)) != 0)
        {
            var region = Native.CreateRoundRectRgn(0, 0, Width + 1, Height + 1,
                CornerRadius, CornerRadius);
            Native.SetWindowRgn(Handle, region, true);
        }
    }

    private void Go(int index)
    {
        _index = Math.Clamp(index, 0, Steps.Length - 1);
        bool isLast = _index == Steps.Length - 1;

        _backButton.Visible = _index > 0;
        _closeButton.Visible = _index == 0;
        _nextButton.Caption = isLast ? "Done" : "Next";
        _nextButton.Left = WindowWidth - Pad - _nextButton.Width;

        Invalidate();
    }

    protected override void OnKeyDown(KeyEventArgs e)
    {
        base.OnKeyDown(e);
        switch (e.KeyCode)
        {
            case Keys.Escape: Close(); break;
            case Keys.Right or Keys.Enter: _nextButton.PerformClick(); break;
            case Keys.Left when _index > 0: Go(_index - 1); break;
        }
    }

    // Borderless, so dragging the body has to stand in for a title bar.
    protected override void OnMouseDown(MouseEventArgs e)
    {
        base.OnMouseDown(e);
        if (e.Button != MouseButtons.Left) return;
        Native.ReleaseCapture();
        Native.SendMessage(Handle, Native.WM_NCLBUTTONDOWN, (IntPtr)Native.HTCAPTION, IntPtr.Zero);
    }

    protected override void OnPaint(PaintEventArgs e)
    {
        var g = e.Graphics;
        g.SmoothingMode = SmoothingMode.AntiAlias;
        g.Clear(Palette.Panel);

        var step = Steps[_index];

        TextRenderer.DrawText(g, "CAFFEINE BAR", MenuFonts.Pill,
            new Point(Pad, Pad), Palette.Blend(Palette.Energy, 0.85),
            TextFormatFlags.NoPadding);

        TextRenderer.DrawText(g, step.Heading, MenuFonts.Heading,
            new Rectangle(Pad, Pad + 20, WindowWidth - Pad * 2, 30),
            Color.White, TextFormatFlags.Left | TextFormatFlags.NoPadding);

        TextRenderer.DrawText(g, step.Body, MenuFonts.Body,
            new Rectangle(Pad, Pad + 56, WindowWidth - Pad * 2, 150),
            Palette.White(0.72),
            TextFormatFlags.Left | TextFormatFlags.WordBreak | TextFormatFlags.NoPadding);

        DrawDots(g);

        using var border = new Pen(Palette.White(0.10), 1);
        using var path = VectorIcons.RoundedRect(
            new RectangleF(0.5f, 0.5f, Width - 1, Height - 1), CornerRadius);
        g.DrawPath(border, path);
    }

    private void DrawDots(Graphics g)
    {
        const int dot = 6;
        const int gap = 7;
        int total = Steps.Length * dot + (Steps.Length - 1) * gap;
        int x = (WindowWidth - total) / 2;
        int y = WindowHeight - Pad - PillButton.ButtonHeight / 2 - dot / 2;

        for (int i = 0; i < Steps.Length; i++)
        {
            using var brush = new SolidBrush(i == _index ? Palette.Energy : Palette.White(0.18));
            g.FillEllipse(brush, x + i * (dot + gap), y, dot, dot);
        }
    }

    /// A small rounded button — filled for the primary action, outlined otherwise.
    private sealed class PillButton : Control
    {
        public const int ButtonHeight = 30;

        private readonly Color _tint;
        private readonly bool _filled;
        private bool _hover;

        public event Action? Clicked;

        public string Caption
        {
            get => Text;
            set { Text = value; UpdateWidth(); Invalidate(); }
        }

        public PillButton(string caption, Color tint, bool filled)
        {
            _tint = tint;
            _filled = filled;
            Height = ButtonHeight;
            Cursor = Cursors.Hand;
            BackColor = Palette.Panel;
            SetStyle(ControlStyles.AllPaintingInWmPaint | ControlStyles.UserPaint |
                     ControlStyles.OptimizedDoubleBuffer | ControlStyles.ResizeRedraw, true);
            Caption = caption;
        }

        public void PerformClick() => Clicked?.Invoke();

        private void UpdateWidth()
        {
            var size = TextRenderer.MeasureText(Text, MenuFonts.Button, Size.Empty,
                TextFormatFlags.NoPadding);
            Width = Math.Max(78, size.Width + 32);
        }

        protected override void OnMouseEnter(EventArgs e) { _hover = true; Invalidate(); }
        protected override void OnMouseLeave(EventArgs e) { _hover = false; Invalidate(); }

        protected override void OnMouseClick(MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Left) Clicked?.Invoke();
        }

        protected override void OnPaint(PaintEventArgs e)
        {
            var g = e.Graphics;
            g.SmoothingMode = SmoothingMode.AntiAlias;
            g.Clear(Palette.Panel);

            var rect = new RectangleF(0.5f, 0.5f, Width - 1, Height - 1);
            using var path = VectorIcons.RoundedRect(rect, ButtonHeight / 2f);

            if (_filled)
            {
                using var fill = new SolidBrush(_hover ? _tint : Palette.Blend(_tint, 0.88));
                g.FillPath(fill, path);
            }
            else
            {
                using var outline = new Pen(Palette.Blend(_tint, _hover ? 0.5 : 0.28), 1);
                g.DrawPath(outline, path);
            }

            var textColor = _filled ? Color.Black : Palette.Blend(_tint, _hover ? 1.0 : 0.8);
            TextRenderer.DrawText(g, Text, MenuFonts.Button, ClientRectangle, textColor,
                TextFormatFlags.HorizontalCenter | TextFormatFlags.VerticalCenter |
                TextFormatFlags.NoPadding);
        }
    }
}

File: Sources/Caffeine/TutorialWindow.swift
import AppKit

/// A short, self-contained walkthrough shown on first launch and re-openable
/// anytime from the icon's right-click menu.
///
/// Clicking, hovering and right-clicking all do different things here, and none
/// of that is discoverable from a menu bar icon on its own — so rather than hope
/// people find it, the app explains itself once and then gets out of the way.
///
/// Unlike CustomMenuWindow this is a real, focusable window: somewhere the user
/// reads and clicks through, not a transient panel that dismisses on the next click.
final class TutorialWindow: NSWindow {
    private struct Step {
        let heading: String
        let body: String
    }

    private static let steps: [Step] = [
        Step(
            heading: "Keep your Mac awake",
            body: """
            Caffeine Bar stops your screen from sleeping for as long as it's active, \
            so your session stays alive. Switch it off and your normal Energy Saver \
            settings take over again. It doesn't change anything permanently.
            """
        ),
        Step(
            heading: "Find the icon",
            body: """
            Caffeine Bar lives in the menu bar, up near the clock.

            If your menu bar is crowded it can get pushed out of sight, which happens \
            easily on a Mac with a notch. Hold Command and drag menu bar icons to \
            reorder them and pull it back where you can see it.
            """
        ),
        Step(
            heading: "Click to toggle",
            body: """
            Click the icon to switch it on and off.

            Filled means it's running: the can fills in and sprouts wings, or the cup \
            puffs steam, depending on the icon style you picked. An empty outline \
            means it's off.
            """
        ),
        Step(
            heading: "Hover for the menu",
            body: """
            Hover the icon to open the full panel. It shows how long you've been \
            active, the energy drink and coffee finders, and the icon style switch.

            Pick Coffee or Energy Drink under Icon Style to change which drink the \
            icon shows.
            """
        ),
        Step(
            heading: "Right-click for settings",
            body: """
            Right-click for Launch at Login, Prevent Sleep on Lid Close, this \
            tutorial, and Quit.

            Closing the lid puts the Mac to sleep in hardware, so that one needs an \
            admin password the first time. Caffeine Bar installs a narrow sudoers rule \
            covering only that command, then never has to ask again.
            """
        ),
    ]

    private var index = 0

    private let headingLabel = NSTextField(labelWithString: "")
    private let bodyLabel = NSTextField(wrappingLabelWithString: "")
    private let dotsView = DotsView(count: TutorialWindow.steps.count)
    private let backButton = PillButton(title: "Back", tint: NSColor.white.withAlphaComponent(0.55), filled: false)
    private let nextButton = PillButton(title: "Next", tint: TutorialWindow.energyTint, filled: true)
    private let skipButton = PillButton(title: "Skip", tint: NSColor.white.withAlphaComponent(0.4), filled: false)

    private static let energyTint = NSColor(calibratedRed: 0.98, green: 0.70, blue: 0.15, alpha: 1)
    private static let windowSize = NSSize(width: 460, height: 320)
    private static let pad: CGFloat = 26

    init() {
        super.init(
            contentRect: NSRect(origin: .zero, size: TutorialWindow.windowSize),
            styleMask: [.borderless],
            backing: .buffered,
            defer: false
        )
        isOpaque = false
        backgroundColor = .clear
        hasShadow = true
        level = .floating
        isMovableByWindowBackground = true
        // Forced dark, to match the custom menu panel rather than following the system.
        appearance = NSAppearance(named: .darkAqua)

        let effect = NSVisualEffectView()
        effect.material = .hudWindow
        effect.state = .active
        effect.blendingMode = .behindWindow
        effect.wantsLayer = true
        effect.layer?.cornerRadius = 16
        effect.layer?.masksToBounds = true
        effect.layer?.borderWidth = 1
        effect.layer?.borderColor = NSColor.white.withAlphaComponent(0.1).cgColor
        contentView = effect

        let tint = NSView()
        tint.wantsLayer = true
        tint.layer?.backgroundColor = NSColor.black.withAlphaComponent(0.35).cgColor
        tint.translatesAutoresizingMaskIntoConstraints = false
        effect.addSubview(tint)

        let brandLabel = NSTextField(labelWithString: "CAFFEINE BAR")
        brandLabel.font = .systemFont(ofSize: 10, weight: .bold)
        brandLabel.textColor = TutorialWindow.energyTint.withAlphaComponent(0.85)
        brandLabel.translatesAutoresizingMaskIntoConstraints = false

        headingLabel.font = .systemFont(ofSize: 21, weight: .bold)
        headingLabel.textColor = .white
        headingLabel.translatesAutoresizingMaskIntoConstraints = false

        bodyLabel.font = .systemFont(ofSize: 13)
        bodyLabel.textColor = NSColor.white.withAlphaComponent(0.72)
        bodyLabel.translatesAutoresizingMaskIntoConstraints = false

        dotsView.translatesAutoresizingMaskIntoConstraints = false

        for view in [brandLabel, headingLabel, bodyLabel, dotsView, backButton, nextButton, skipButton] {
            effect.addSubview(view)
        }

        backButton.onClick = { [weak self] in self?.go(to: (self?.index ?? 1) - 1) }
        skipButton.onClick = { [weak self] in self?.finish() }
        nextButton.onClick = { [weak self] in
            guard let self else { return }
            if index == TutorialWindow.steps.count - 1 { finish() } else { go(to: index + 1) }
        }

        let pad = TutorialWindow.pad
        NSLayoutConstraint.activate([
            tint.topAnchor.constraint(equalTo: effect.topAnchor),
            tint.bottomAnchor.constraint(equalTo: effect.bottomAnchor),
            tint.leadingAnchor.constraint(equalTo: effect.leadingAnchor),
            tint.trailingAnchor.constraint(equalTo: effect.trailingAnchor),

            brandLabel.topAnchor.constraint(equalTo: effect.topAnchor, constant: pad),
            brandLabel.leadingAnchor.constraint(equalTo: effect.leadingAnchor, constant: pad),

            headingLabel.topAnchor.constraint(equalTo: brandLabel.bottomAnchor, constant: 8),
            headingLabel.leadingAnchor.constraint(equalTo: effect.leadingAnchor, constant: pad),
            headingLabel.trailingAnchor.constraint(equalTo: effect.trailingAnchor, constant: -pad),

            bodyLabel.topAnchor.constraint(equalTo: headingLabel.bottomAnchor, constant: 12),
            bodyLabel.leadingAnchor.constraint(equalTo: effect.leadingAnchor, constant: pad),
            bodyLabel.trailingAnchor.constraint(equalTo: effect.trailingAnchor, constant: -pad),

            backButton.leadingAnchor.constraint(equalTo: effect.leadingAnchor, constant: pad),
            backButton.bottomAnchor.constraint(equalTo: effect.bottomAnchor, constant: -pad),
            skipButton.leadingAnchor.constraint(equalTo: effect.leadingAnchor, constant: pad),
            skipButton.bottomAnchor.constraint(equalTo: effect.bottomAnchor, constant: -pad),
            nextButton.trailingAnchor.constraint(equalTo: effect.trailingAnchor, constant: -pad),
            nextButton.bottomAnchor.constraint(equalTo: effect.bottomAnchor, constant: -pad),

            dotsView.centerXAnchor.constraint(equalTo: effect.centerXAnchor),
            dotsView.centerYAnchor.constraint(equalTo: nextButton.centerYAnchor),
        ])

        go(to: 0)
    }

    override var canBecomeKey: Bool { true }

    /// Centers on whichever screen the pointer is on, so it lands where the user
    /// just clicked rather than always on the primary display.
    func present() {
        let screen = NSScreen.screens.first { NSMouseInRect(NSEvent.mouseLocation, $0.frame, false) }
            ?? NSScreen.main
        if let visible = screen?.visibleFrame {
            setFrameOrigin(NSPoint(
                x: visible.midX - frame.width / 2,
                y: visible.midY - frame.height / 2
            ))
        }
        NSApp.activate(ignoringOtherApps: true)
        makeKeyAndOrderFront(nil)
    }

    private func go(to newIndex: Int) {
        index = min(max(newIndex, 0), TutorialWindow.steps.count - 1)
        let step = TutorialWindow.steps[index]
        let isLast = index == TutorialWindow.steps.count - 1

        headingLabel.stringValue = step.heading
        bodyLabel.stringValue = step.body
        backButton.isHidden = index == 0
        skipButton.isHidden = index != 0
        nextButton.title = isLast ? "Done" : "Next"
        dotsView.current = index
    }

    private func finish() {
        orderOut(nil)
    }

    override func keyDown(with event: NSEvent) {
        switch event.keyCode {
        case 53: finish()                                   // Escape
        case 124, 36: nextButton.onClick?()                 // Right arrow, Return
        case 123 where index > 0: go(to: index - 1)         // Left arrow
        default: super.keyDown(with: event)
        }
    }

    /// The step indicator: one filled dot for the current step, dim for the rest.
    private final class DotsView: NSView {
        private let count: Int
        var current = 0 { didSet { needsDisplay = true } }

        init(count: Int) {
            self.count = count
            super.init(frame: .zero)
            let dot: CGFloat = 6, gap: CGFloat = 7
            widthAnchor.constraint(equalToConstant: CGFloat(count) * dot + CGFloat(count - 1) * gap).isActive = true
            heightAnchor.constraint(equalToConstant: dot).isActive = true
        }

        required init?(coder: NSCoder) { fatalError("not supported") }

        override func draw(_ dirtyRect: NSRect) {
            let dot: CGFloat = 6, gap: CGFloat = 7
            for i in 0..<count {
                let color = i == current ? TutorialWindow.energyTint : NSColor.white.withAlphaComponent(0.18)
                color.setFill()
                NSBezierPath(ovalIn: NSRect(x: CGFloat(i) * (dot + gap), y: 0, width: dot, height: dot)).fill()
            }
        }
    }

    /// A small rounded button — filled for the primary action, outlined otherwise.
    private final class PillButton: NSView {
        private let tint: NSColor
        private let filled: Bool
        private var isHovered = false { didSet { needsDisplay = true } }
        private var trackingAreaRef: NSTrackingArea?
        private var widthConstraint: NSLayoutConstraint?

        var onClick: (() -> Void)?

        var title: String {
            didSet {
                widthConstraint?.constant = PillButton.width(for: title)
                needsDisplay = true
            }
        }

        init(title: String, tint: NSColor, filled: Bool) {
            self.title = title
            self.tint = tint
            self.filled = filled
            super.init(frame: .zero)
            translatesAutoresizingMaskIntoConstraints = false
            heightAnchor.constraint(equalToConstant: 30).isActive = true
            let width = widthAnchor.constraint(equalToConstant: PillButton.width(for: title))
            width.isActive = true
            widthConstraint = width
        }

        required init?(coder: NSCoder) { fatalError("not supported") }

        private static let font = NSFont.systemFont(ofSize: 12, weight: .semibold)

        private static func width(for title: String) -> CGFloat {
            let size = (title as NSString).size(withAttributes: [.font: font])
            return max(78, ceil(size.width) + 32)
        }

        override func updateTrackingAreas() {
            super.updateTrackingAreas()
            if let trackingAreaRef { removeTrackingArea(trackingAreaRef) }
            let area = NSTrackingArea(
                rect: bounds,
                options: [.mouseEnteredAndExited, .activeAlways, .inVisibleRect],
                owner: self,
                userInfo: nil
            )
            addTrackingArea(area)
            trackingAreaRef = area
        }

        override func mouseEntered(with event: NSEvent) { isHovered = true }
        override func mouseExited(with event: NSEvent) { isHovered = false }
        override func mouseDown(with event: NSEvent) { onClick?() }

        override func resetCursorRects() {
            addCursorRect(bounds, cursor: .pointingHand)
        }

        override func draw(_ dirtyRect: NSRect) {
            let rect = bounds.insetBy(dx: 0.5, dy: 0.5)
            let path = NSBezierPath(roundedRect: rect, xRadius: rect.height / 2, yRadius: rect.height / 2)

            if filled {
                (isHovered ? tint : tint.withAlphaComponent(0.88)).setFill()
                path.fill()
            } else {
                tint.withAlphaComponent(isHovered ? 0.5 : 0.28).setStroke()
                path.lineWidth = 1
                path.stroke()
            }

            let color: NSColor = filled ? .black : tint.withAlphaComponent(isHovered ? 1.0 : 0.8)
            let attributes: [NSAttributedString.Key: Any] = [
                .font: PillButton.font,
                .foregroundColor: color,
            ]
            let size = (title as NSString).size(withAttributes: attributes)
            (title as NSString).draw(
                at: NSPoint(x: bounds.midX - size.width / 2, y: bounds.midY - size.height / 2),
                withAttributes: attributes
            )
        }
    }
}
